src/transformers/models/t5/modeling_t5.py

```diff
@@ -621,7 +621,7 @@ def forward(
     ):
 
         if past_key_value is not None:
-            assert self.is_decoder, "Only decoder can use `past_key_values`"
+            # assert self.is_decoder, "Only decoder can use `past_key_values`"
```

Comment:
Thanks for the review! 
It's true that `past_key_values`  is never used in the encoder. And this is already ensured by the `forward()` in T5ForConditionalGeneration:
https://github.com/huggingface/transformers/blob/9e5fa33cdc5fb596bdf08b88aadf9e0b551a5b76/src/transformers/models/t5/modeling_t5.py#L1571-L1579
So, I think it's unnecessary to double-check this in T5Block, which is the case in BART.

As for the purpose of using `past_key_values `in the encoder, some research uses it to enrich the context for the input sequence. And according to my experiments in BART model, this actually works.
And in BARTForConditionalGeneration, though the `past_key_values` is not used in the encoder, it can be done by explicitly calling the `BARTEncoder(past_key_values=xxx,` **kwargs), and then pass the `encoder_outputs` to BART.

Therefore, I hope we can remove this `assert` in the T5 model as well.

By the way, if this `assert` is deleted, we should also change the following code to `past_key_values[-2:]` to avoid IndexError in the encoder because the length of  `past_key_values` ' is 2 for self-attention.
https://github.com/huggingface/transformers/blob/7db2a79b387fd862ffb0af72f7148e6371339c7f/src/transformers/models/t5/modeling_t5.py#L635

Reply:
I'm with @patil-suraj here - I think this PR is about enabling to use `past_key_values` for the T5 decoder in training - this line removal does not have anything to do with this IMO. Can we maybe leave this for a potential second PR?

Reply:
Sure! Let's fix the decoder first.